bluepyentity/register.py

```diff
@@ -0,0 +1,472 @@
+
+        # Fields to not store in Nexus
+        refined.pop("brainRegion", None)
+
+        return self._with_defaults(refined)
+
+    @property
+    def possible_ids(self):
+        """Return the fields that possibly contains nexus ids"""
+        return self._schema_definition["ids"]
+
+    @property
+    def resource(self):
+        """Access the resource."""
+        return self._resource
+
+    @property
+    def schema(self):
+        """Schema for the entity."""
+        # TODO: figure out how to get this from Nexus and use it to validate types etc.
+        return self._schema_definition["schema"]
+
+    @property
+    def type(self):
+        """Return the Nexus type of the resource."""
+        return self._definition["type"]
+
+    @cached_property
+    def _schema_definition(self):
+        return get_entity_definition(self.type)
+
+    def register(self):
+        """Register the resource in Nexus."""
+        if existing := self._find_existing():
+            bucket = self._forge._store.bucket  # pylint: disable=protected-access
+            raise RuntimeError(
+                f"Similar '{self.type}' definition already exists in project "
+                f"'{bucket}' with an id of: '{existing.id}'"
+            )
+
+        try:
+            schema_id = self._forge._model.schema_id(self.type)  # pylint: disable=protected-access
+        except ValueError:
+            schema_id = None
+
+        try:
+            self._forge.register(self.resource, schema_id=schema_id)
+        except Exception as err:
+            raise RuntimeError(err) from err
+
+    def _check_if_valid(self):
+        self._check_required()
+        # NOTE: should this be removed, if schemas are handled by the backend?
+        # Should extra attributes be allowed? Currently, the backend allows them.
+        # self._check_schema()
+
+    def _check_required(self):
+        """Check that the required items are defined."""
+        missing = self.required - set(self.definition)
+        assert not missing, f"Missing attributes: {missing}"
+
+    def _check_schema(self):
+        disallowed = set(self.definition) - set(self.schema)
+        disallowed -= {"type"}
+        assert not disallowed, f"Disallowed attributes: {disallowed}"
+
+    def _create_resource(self):
+        """Create a kgforge Resource of the definition."""
+        definition = self.definition
+        definition["distribution"] = self._get_distribution()
+
+        return kgforge.core.Resource.from_json(definition)
+
+    def _format_attributes(self, definition):
+        """Format needed values in the definition."""
+        return definition
+
+    def _find_existing(self):
+        """Find resource matching the definition in Nexus."""
+        found = self._forge.search({"type": self.type}, cross_bucket=True)
+        for r in found:
+            if self._is_equal_to_nexus_resource(r):
+                return r
+
+        return None
+
+    def _get_distribution(self):
```

Comment:
Minor: Wouldn't a plural be more appropriate here if it returns a list?

Reply:
I would tend to agree. However, despite being a list of items, the nexus field is defined in singular form: `distribution` so I went with that.